internal/providers/oracleoci/oracleoci.go

```diff
@@ -0,0 +1,64 @@
+// Copyright 2017 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package oracleoci
+
+import (
+	"encoding/json"
+	"strings"
+	"time"
+
+	"github.com/coreos/coreos-metadata/internal/providers"
+	"github.com/coreos/coreos-metadata/internal/retry"
+)
+
+type instanceData struct {
+	AvailabilityDomain string   `json:"availabilityDomain"`
+	CompartmentId      string   `json:"compartmentId"`
+	DisplayName        string   `json:"displayName"`
+	Id                 string   `json:"id"`
+	Image              string   `json:"image"`
+	Region             string   `json:"region"`
+	Shape              string   `json:"shape"`
+	TimeCreated        uint64   `json:"timeCreated"`
+	Metadata           metadata `json:"metadata"`
+}
+
+type metadata struct {
+	SshAuthorizedKeys string `json:"ssh_authorized_keys"`
+}
+
+func FetchMetadata() (providers.Metadata, error) {
+	instanceDataBlob, err := retry.Client{
+		InitialBackoff: time.Second,
+		MaxBackoff:     time.Second * 5,
+		MaxAttempts:    10,
+	}.Get("http://169.254.169.254/opc/v1/instance/")
+	if err != nil {
+		return providers.Metadata{}, err
+	}
+
+	var data instanceData
+	err = json.Unmarshal(instanceDataBlob, &data)
+	if err != nil {
+		return providers.Metadata{}, err
+	}
+	return providers.Metadata{
+		Attributes: map[string]string{
+			"ORACLE_OCI_HOSTNAME": data.DisplayName,
+		},
+		Hostname: data.DisplayName,
```

Comment:
OCI provides separate fields for the display name and the hostname.  (`scripts/oem/oracle/launch-instance` currently sets both to the same value.)  It appears the only way to get the hostname is via reverse lookup of the private IP.  There can be a separate hostname for each VNIC, but we're probably only interested in the first one.

Reply:
In particular, the display name can be up to 255 characters and can contain e.g. spaces.